### PR TITLE
Implementation of array:sort

### PR DESCRIPTION
**Pull request by adamretter**
Back-ported from FusionDB.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/value/AtomicValueComparator.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            return -1;
+        } else if (o2 == null) {
+            return 1;
+        }
+
+        try {
+            return o1.compareTo(collator, o2);
+        } catch (final XPathException e) {
+            LOG.error(e.getMessage(), e);
+            throw new ClassCastException(e.getMessage());
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/value/AtomicValueComparator.java:77]: ??

File: exist-core/src/test/xquery/arrays/arrays.xql [hunk 2]
```diff
@@ -1009,3 +1009,27 @@ declare
 function arr:put_pos_illegal_index() {
     array:put(["a", "b", "c", "d"], 5, "x")?*
 };
+
+declare
+    %test:assertEquals(1, 3, 4, 5, 6)
+function arr:sort-int() {
```
Comment (dizzzz) [inline exist-core/src/test/xquery/arrays/arrays.xql:1015]: what about a test doing 'mixed' content? ints, strings, decimals

**Review by dizzzz — APPROVED** [review]

File: exist-core/src/main/java/org/exist/xquery/value/AtomicValueComparator.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/value/AtomicValueComparator.java:77]: @dizzzz What other option do we have? The API says we can throw either:

```
     * @throws NullPointerException if an argument is null and this
     *         comparator does not permit null arguments
     * @throws ClassCastException if the arguments' types prevent them from
     *         being compared by this comparator.
```

So `ClassCastException` seemed like the *most* correct option to me.

File: exist-core/src/test/xquery/arrays/arrays.xql (hunk 2 above)
Comment (adamretter) [reply exist-core/src/test/xquery/arrays/arrays.xql:1015]: Done.

**Comment by joewiz** [comment]
@adamretter Thank you for helping eXist move toward completing eXist's coverage of the XQuery 3.1 spec!

**Pull request merged**